Create a Python repository fixture

asmacdo opened the pull request.

Ichimonji10 commented on python/gen-fixtures.sh:
```diff
@@ -0,0 +1,42 @@
+# Copy static files of Python repository into fixtures directory
+#
+set -euo pipefail
+readonly image_name='busybox:latest'
```
Unused variable.

Ichimonji10 commented on python/gen-fixtures.sh:
```diff
@@ -0,0 +1,42 @@
+cat <<EOF
+<output_dir>
+    The directory into which generated fixtures are placed.
+EOF
```
Please add a brief description of `<assets_dir>` here.

By the way, the `fmt` on my system is smart enough to handle simple definition lists. I may have been overly-cautious when deciding to use both `fmt` and `cat`. This input:

```sh
fmt -u <<EOF
usage: foo bar

Description of command. Asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf.

definition term 1
    Definition data for definition term 1. Asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf.
definition term 2
    Foo bar biz baz.
EOF
```

Gives this output:

```
usage: foo bar

Description of command. Asdf asdf asdf asdf asdf asdf asdf asdf asdf
asdf asdf asdf asdf asdf asdf asdf.

definition term 1
    Definition data for definition term 1. Asdf asdf asdf asdf asdf asdf
    asdf asdf asdf asdf asdf.
definition term 2
    Foo bar biz baz.
```

Ichimonji10 commented:
This looks great! I'm going serve up these fixtures locally and have Pulp sync them.

Ichimonji10 commented on python/gen-fixtures.sh:
```diff
@@ -0,0 +1,42 @@
+
+# Create output_dir and generate fixture data.
+mkdir "${output_dir}"
+cp -r "${assets_dir}" "${output_dir}"
```
This causes the entire "assets" and "test_py" directories to be copied to the `fixtures/python`, producing:

```
fixtures/
└── python
    └── assets
        └── test_py
            ├── packages
            │   └── source
            │       └── s
            │           └── shelf-reader
            │               ├── shelf_reader-0.1-py2-none-any.whl
            │               └── shelf-reader-0.1.tar.gz
            ├── pypi
            │   └── shelf-reader
            │       └── json
            │           └── index.html
            └── simple
                ├── index.html
                └── shelf-reader
                    └── index.html
```

Consider dropping the `test_py` directory and writing this instead:

```sh
cp -rt "${output_dir}" "${assets_dir}"/*
```

Ichimonji10 commented:
I generated these files and served them up with a dumb HTTP file server:

```sh
make fixtures/python
cd fixtures/python/assets/test_py/
python3 -m http.server
```

On the same machine, I then created a repo and synced it:

```sh
pulp-admin python repo create --repo-id ichi10 --feed http://localhost:8000 --package-names shelf-reader
pulp-admin python repo sync run --repo-id ichi10
pulp-admin python repo list --repo-id ichi10
pulp-admin python repo delete --repo-id ichi10
```

The "Downloading and processing Python packages." step failed. (Unsurprisingly, `pulp-admin` proceeded.) This error was visible with `journalctl -u '*pulp*'`:

```
Jun 08 14:39:11 pulp.example.com pulp[6653]: nectar.downloaders.threaded:ERROR: (6653-88096) Invalid URL '../../../packages/source/s/shelf-reader/shelf-reader-0.1.tar.gz#md5=2dac570a33d88ca224be86759be59376': No schema supplied. Perhaps you meant http://../../../packages/source/s/shelf-reader/shelf-reader-0.1.tar.gz#md5=2dac570a33d88ca224be86759be59376?
Jun 08 14:39:11 pulp.example.com pulp[6653]: nectar.downloaders.threaded:ERROR: (6653-88096) Traceback (most recent call last):
Jun 08 14:39:11 pulp.example.com pulp[6653]: nectar.downloaders.threaded:ERROR: (6653-88096)   File "/usr/lib/python2.7/site-packages/nectar/downloaders/threaded.py", line 221, in _fetch
Jun 08 14:39:11 pulp.example.com pulp[6653]: nectar.downloaders.threaded:ERROR: (6653-88096)     self.config.read_timeout))
Jun 08 14:39:11 pulp.example.com pulp[6653]: nectar.downloaders.threaded:ERROR: (6653-88096)   File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 487, in get
Jun 08 14:39:11 pulp.example.com pulp[6653]: nectar.downloaders.threaded:ERROR: (6653-88096)     return self.request('GET', url, **kwargs)
Jun 08 14:39:11 pulp.example.com pulp[6653]: nectar.downloaders.threaded:ERROR: (6653-88096)   File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 461, in request
Jun 08 14:39:11 pulp.example.com pulp[6653]: nectar.downloaders.threaded:ERROR: (6653-88096)     prep = self.prepare_request(req)
Jun 08 14:39:11 pulp.example.com pulp[6653]: nectar.downloaders.threaded:ERROR: (6653-88096)   File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 394, in prepare_request
Jun 08 14:39:11 pulp.example.com pulp[6653]: nectar.downloaders.threaded:ERROR: (6653-88096)     hooks=merge_hooks(request.hooks, self.hooks),
Jun 08 14:39:11 pulp.example.com pulp[6653]: nectar.downloaders.threaded:ERROR: (6653-88096)   File "/usr/lib/python2.7/site-packages/requests/models.py", line 295, in prepare
Jun 08 14:39:11 pulp.example.com pulp[6653]: nectar.downloaders.threaded:ERROR: (6653-88096)     self.prepare_url(url, params)
Jun 08 14:39:11 pulp.example.com pulp[6653]: nectar.downloaders.threaded:ERROR: (6653-88096)   File "/usr/lib/python2.7/site-packages/requests/models.py", line 355, in prepare_url
Jun 08 14:39:11 pulp.example.com pulp[6653]: nectar.downloaders.threaded:ERROR: (6653-88096)     raise MissingSchema(error)
Jun 08 14:39:11 pulp.example.com pulp[6653]: nectar.downloaders.threaded:ERROR: (6653-88096) MissingSchema: Invalid URL '../../../packages/source/s/shelf-reader/shelf-reader-0.1.tar.gz#md5=2dac570a33d88ca224be86759be59376': No schema supplied. Perhaps you meant http://../../../packages/source/s/shelf-reader/shelf-reader-0.1.tar.gz#md5=2dac570a33d88ca224be86759be59376?
```

Please fix.

asmacdo commented:
Fixed the little things, copied from `assets/*` rather than copying `assets/`, and renamed `test_py` to `example_repo`.

This will not work with current master of pulp_python. This is prep work for merging a feature branch.

asmacdo commented on python/assets/example_repo/pypi/shelf-reader/json/index.json:
```diff
@@ -0,0 +1 @@
+{"info": {"author": "Austin Macdonald", "name": "shelf-reader", "summary": "Make sure your collections are in call number order."}, "releases": {"0.1": [{"path": "../../../source/s/shelf-reader/shelf-reader-0.1.tar.gz#md5=2dac570a33d88ca224be86759be59376", "md5_digest": "2dac570a33d88ca224be86759be59376", "packagetype": "sdist", "filename": "shelf-reader-0.1.tar.gz"}, {"path": "../../../source/s/ ...
```
@Ichimonji10, I changed this from index.html to index.json to make sure that the content type is correct. Whatever server is hosting the assets will need to forward a directory request to index.json as well as index.html.

elyezer commented:
Is there a way to test this? That will help merging this.

Ichimonji10 commented:
Merged as 8922c8c5cd0192a85bd951ae80657ac9c4bcdf24, with a follow-up commit of 3bc1136ad54b232c9d15dc4ceafa045ec576d656. Thank you, @asmacdo!

I've tested these changes by editing Pulp Smash's `pulp_smash.constants.PULP_*` constants, and then running both the existing tests and https://github.com/PulpQE/pulp-smash/pull/318.
